Auth0ContextInterface has incorrect interface typing

### Checklist

- [x] The issue can be reproduced in the [react-native-auth0 sample app](https://github.com/auth0-samples/auth0-react-native-sample) (or N/A).
- [x] I have looked into the [Readme](https://github.com/auth0/react-native-auth0#readme), [Examples](https://github.com/auth0/react-native-auth0/blob/master/EXAMPLES.md), and [FAQ](https://github.com/auth0/react-native-auth0/blob/master/FAQ.md) and have not found a suitable solution or answer.
- [x] I have looked into the [API documentation](https://auth0.github.io/react-native-auth0/) and have not found a suitable solution or answer.
- [x] I have searched the [issues](https://github.com/auth0/react-native-auth0/issues) and have not found a suitable solution or answer.
- [x] I have searched the [Auth0 Community](https://community.auth0.com) forums and have not found a suitable solution or answer.
- [x] I agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Description

Some of the types on `Auth0ContextInterface` don't seem to match the implementation as well as the examples. This is triggering lint messages for https://typescript-eslint.io/rules/unbound-method/ when used following examples.

Take `getCredentials` as an example: https://github.com/auth0/react-native-auth0/blob/92b3637023275bd59590cea96ecf886c47306bfb/src/hooks/Auth0Context.ts#L75-L80

The way that function's type is written implies that getCredentials is a member of a class that makes use of `this` internally. Because of that, writing code like `const { getCredentials } = useAuth0(); getCredentials();` will fail because the `this` parameter isn't being set properly. Example of where that lint would fail: https://github.com/auth0/react-native-auth0/blob/92b3637023275bd59590cea96ecf886c47306bfb/example/src/screens/hooks/Api.tsx#L15

Contrast that with the type for `clearCredentials`: https://github.com/auth0/react-native-auth0/blob/92b3637023275bd59590cea96ecf886c47306bfb/src/hooks/Auth0Context.ts#L89

That is saying that the property is a function which doesn't need a bound `this` and will work with a destructuring assignment.

Looking at the implementation, many of the properties of the `Auth0ContextInterface` interface likely should be written similar to `clearCredentials`. In the case of `getCredentials` it would likely look something like the following. Note the `:` after the name and the change to `=>` in the type specification:

```
getCredentials: (
  scope?: string,
  minTtl?: number,
  parameters?: Record<string, unknown>,
  forceRefresh?: boolean
) => Promise<Credentials>;
```

### Reproduction

* Configure typescript-eslint (8.44.1) with the referenced lint enabled with the default settings.
* Run over the linter over the examples in the repo.

### Additional context

_No response_

### react-native-auth0 version

5.0.0-beta.6

### React Native version

0.79.2

### Expo version

53.0.22

### Platform

iOS

### Platform version(s)

All

> Is this issue specifically related to the linter?

Bear with me as the response is a little nuanced. The linter was calling out that the way in which we were making use of the `getCredentials` function is incorrect based on the type of the function. Specifically, the following pattern is being flagged as problematic:

```typescript
const { getCredentials } = useAuth0();
/*...*/
getCredentials(/*...*/); // NOTE: `this` is `undefined`
```

The type for `getCredentials`, as written, specifies that `this` should be bound to an instance. Specifically, the linter is saying that the following is how to call `getCredentials`:

```typescript
const auth0 = useAuth0();
/*...*/
auth0.getCredentials(/*...*/); // NOTE: `this` is bound to the `auth0` instance.
```

Thankfully, `getCredentials` doesn't actually use `this`, so there is no behavioral differences between the two versions.

The only reason I'm pointing this out is that the examples in this repository are often of the first form. It seems like the intention is to encourage and allow the first form. To make the examples and the types consistent, the way in which the type of `getCredentials` is written should change slightly.

> Is it observed with any particular version typescript-eslint (8.44.1)?

8.44.1 is the version that we were using when I first discovered the error. I've not attempted to bisect the version to see if the behavior has changed.

---

I whipped together a branch which demonstrates the issue at hand: https://github.com/nairb774/react-native-auth0/tree/typescript-eslint-example

```
cd react-native-auth0/example
yarn eslint src/screens/hooks/Api.tsx

react-native-auth0/example/src/screens/hooks/Api.tsx
  15:17  error  A method that is not declared with `this: void` may cause unintentional scoping of `this` when separated from its object.
Consider using an arrow function or explicitly `.bind()`ing the method to avoid calling the method with an unintended `this` value.
If a function does not access `this`, it can be annotated with `this: void`  @typescript-eslint/unbound-method
```

See the commit message for more details: https://github.com/nairb774/react-native-auth0/commit/0d93dd3f4b10a31306dfccadb1ce966719fe6da8

--- a/eslint.config.mjs
+++ b/eslint.config.mjs
@@ -2,6 +2,7 @@ import { fixupConfigRules } from '@eslint/compat';
 import { FlatCompat } from '@eslint/eslintrc';
 import js from '@eslint/js';
 import prettier from 'eslint-plugin-prettier';
+import tseslint from 'typescript-eslint';
 import { defineConfig } from 'eslint/config';
 import path from 'node:path';
 import { fileURLToPath } from 'node:url';
@@ -30,6 +31,23 @@ export default defineConfig([
           useTabs: false,
         },
       ],
+    },
+  },
+  // TypeScript-specific configuration for type-checked rules
+  {
+    files: ['**/*.ts', '**/*.tsx'],
+    ignores: ['**/__tests__/**', '**/__mocks__/**', '**/*.spec.ts', '**/*.spec.tsx', '**/*.test.ts', '**/*.test.tsx'],
+    languageOptions: {
+      parserOptions: {
+        project: true,
+        tsconfigRootDir: import.meta.dirname,
+      },
+    },
+    rules: {
+      // Enable unbound-method rule to catch interface typing issues
+      // This helps ensure methods in interfaces use arrow function syntax
+      // instead of method syntax when they don't use 'this'
+      '@typescript-eslint/unbound-method': 'error',
     },
   },
   {

--- a/package.json
+++ b/package.json
@@ -132,7 +132,8 @@
     "typedoc": "^0.28.2",
     "typedoc-plugin-missing-exports": "^4.0.0",
     "typedoc-plugin-replace-text": "^4.2.0",
-    "typescript": "5.2.2"
+    "typescript": "5.2.2",
+    "typescript-eslint": "^8.46.2"
   },
   "dependencies": {
     "@auth0/auth0-spa-js": "2.7.0",

--- a/src/hooks/Auth0Context.ts
+++ b/src/hooks/Auth0Context.ts
@@ -39,10 +39,10 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the user's credentials upon successful authentication.
    * @throws {AuthError} If the authentication fails.
    */
-  authorize(
+  authorize: (
     parameters?: WebAuthorizeParameters,
     options?: NativeAuthorizeOptions
-  ): Promise<Credentials>;
+  ) => Promise<Credentials>;
 
   /**
    * Clears the user's session and logs them out.
@@ -51,10 +51,10 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves when the session has been cleared.
    * @throws {AuthError} If the logout fails.
    */
-  clearSession(
+  clearSession: (
     parameters?: ClearSessionParameters,
     options?: NativeClearSessionOptions
-  ): Promise<void>;
+  ) => Promise<void>;
 
   /**
    * Saves the user's credentials.
@@ -62,7 +62,7 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves when the credentials have been saved.
    * @throws {AuthError} If the save fails.
    */
-  saveCredentials(credentials: Credentials): Promise<void>;
+  saveCredentials: (credentials: Credentials) => Promise<void>;
 
   /**
    * Retrieves the stored credentials, refreshing them if necessary.
@@ -73,12 +73,12 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the user's credentials.
    * @throws {AuthError} If credentials cannot be retrieved or refreshed.
    */
-  getCredentials(
+  getCredentials: (
     scope?: string,
     minTtl?: number,
     parameters?: Record<string, unknown>,
     forceRefresh?: boolean
-  ): Promise<Credentials>;
+  ) => Promise<Credentials>;
 
   /**
    * Clears the user's credentials without clearing their web session and logs them out.
@@ -96,13 +96,13 @@ export interface Auth0ContextInterface extends AuthState {
    * @param minTtl The minimum time-to-live (in seconds) required for the access token to be considered valid. Defaults to 0.
    * @returns A promise that resolves with `true` if valid credentials exist, `false` otherwise.
    */
-  hasValidCredentials(minTtl?: number): Promise<boolean>;
+  hasValidCredentials: (minTtl?: number) => Promise<boolean>;
 
   /**
    * Cancels the ongoing web authentication process.
    * This works only on iOS. On other platforms, it will resolve without performing an action.
    */
-  cancelWebAuth(): Promise<void>;
+  cancelWebAuth: () => Promise<void>;
 
   /**
    * Authenticates a user with their username and password.
@@ -111,9 +111,9 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the user's credentials.
    * @throws {AuthError} If the authentication fails.
    */
-  loginWithPasswordRealm(
+  loginWithPasswordRealm: (
     parameters: PasswordRealmParameters
-  ): Promise<Credentials>;
+  ) => Promise<Credentials>;
 
   /**
    * Creates a new user in a database connection.
@@ -121,7 +121,7 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the new user's profile information.
    * @throws {AuthError} If the user creation fails.
    */
-  createUser(parameters: CreateUserParameters): Promise<Partial<User>>;
+  createUser: (parameters: CreateUserParameters) => Promise<Partial<User>>;
 
   /**
    * Resets the user's password.
@@ -129,7 +129,7 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves when the password has been reset.
    * @throws {AuthError} If the reset fails.
    */
-  resetPassword(parameters: ResetPasswordParameters): Promise<void>;
+  resetPassword: (parameters: ResetPasswordParameters) => Promise<void>;
 
   /**
    * Exchanges an authorization code for tokens.
@@ -138,7 +138,9 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the user's credentials.
    * @throws {AuthError} If the exchange fails.
    */
-  authorizeWithExchange(parameters: ExchangeParameters): Promise<Credentials>;
+  authorizeWithExchange: (
+    parameters: ExchangeParameters
+  ) => Promise<Credentials>;
 
   /**
    * Exchanges an authorization code for native social tokens.
@@ -146,16 +148,16 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the user's credentials.
    * @throws {AuthError} If the exchange fails.
    */
-  authorizeWithExchangeNativeSocial(
+  authorizeWithExchangeNativeSocial: (
     parameters: ExchangeNativeSocialParameters
-  ): Promise<Credentials>;
+  ) => Promise<Credentials>;
 
   /**
    * Sends a verification code to the user's email.
    * @param parameters The parameters for sending the email code.
    * @throws {AuthError} If sending the email code fails.
    */
-  sendEmailCode(parameters: PasswordlessEmailParameters): Promise<void>;
+  sendEmailCode: (parameters: PasswordlessEmailParameters) => Promise<void>;
 
   /**
    * Authorizes a user with their email.
@@ -163,7 +165,9 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the user's credentials.
    * @throws {AuthError} If the authorization fails.
    */
-  authorizeWithEmail(parameters: LoginEmailParameters): Promise<Credentials>;
+  authorizeWithEmail: (
+    parameters: LoginEmailParameters
+  ) => Promise<Credentials>;
 
   /**
    /**
@@ -171,7 +175,7 @@ export interface Auth0ContextInterface extends AuthState {
    * @param parameters The parameters for sending the SMS code.
    * @throws {AuthError} If sending the SMS code fails.
    */
-  sendSMSCode(parameters: PasswordlessSmsParameters): Promise<void>;
+  sendSMSCode: (parameters: PasswordlessSmsParameters) => Promise<void>;
 
   /**
    * Authorizes a user with their SMS.
@@ -179,7 +183,7 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the user's credentials.
    * @throws {AuthError} If the authorization fails.
    */
-  authorizeWithSMS(parameters: LoginSmsParameters): Promise<Credentials>;
+  authorizeWithSMS: (parameters: LoginSmsParameters) => Promise<Credentials>;
 
   /**
    * Sends a multifactor challenge to the user.
@@ -187,9 +191,9 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves when the challenge has been sent.
    * @throws {AuthError} If sending the challenge fails.
    */
-  sendMultifactorChallenge(
+  sendMultifactorChallenge: (
     parameters: MfaChallengeParameters
-  ): Promise<MfaChallengeResponse>;
+  ) => Promise<MfaChallengeResponse>;
 
   /**
    * Authorizes a user with out-of-band (OOB) authentication.
@@ -197,7 +201,7 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the user's credentials.
    * @throws {AuthError} If the authorization fails.
    */
-  authorizeWithOOB(parameters: LoginOobParameters): Promise<Credentials>;
+  authorizeWithOOB: (parameters: LoginOobParameters) => Promise<Credentials>;
 
   /**
    * Authorizes a user with a one-time password (OTP).
@@ -205,7 +209,7 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the user's credentials.
    * @throws {AuthError} If the authorization fails.
    */
-  authorizeWithOTP(parameters: LoginOtpParameters): Promise<Credentials>;
+  authorizeWithOTP: (parameters: LoginOtpParameters) => Promise<Credentials>;
 
   /**
    * Authorizes a user with a recovery code.
@@ -213,12 +217,12 @@ export interface Auth0ContextInterface extends AuthState {
    * @returns A promise that resolves with the user's credentials.
    * @throws {AuthError} If the authorization fails.
    */
-  authorizeWithRecoveryCode(
+  authorizeWithRecoveryCode: (
     parameters: LoginRecoveryCodeParameters
-  ): Promise<Credentials>;
+  ) => Promise<Credentials>;
 
   // Token Management
-  revokeRefreshToken(parameters: RevokeOptions): Promise<void>;
+  revokeRefreshToken: (parameters: RevokeOptions) => Promise<void>;
 
   /**
    * Generates DPoP headers for making authenticated requests to custom APIs.
@@ -244,7 +248,9 @@ export interface Auth0ContextInterface extends AuthState {
    * }
    * ```
    */
-  getDPoPHeaders(params: DPoPHeadersParams): Promise<Record<string, string>>;
+  getDPoPHeaders: (
+    params: DPoPHeadersParams
+  ) => Promise<Record<string, string>>;
 }
 
 const stub = (): any => {

--- a/src/platforms/native/bridge/NativeBridgeManager.ts
+++ b/src/platforms/native/bridge/NativeBridgeManager.ts
@@ -40,7 +40,9 @@ export class NativeBridgeManager implements INativeBridge {
 
   async hasValidInstance(clientId: string, domain: string): Promise<boolean> {
     return this.a0_call(
-      Auth0NativeModule.hasValidAuth0InstanceWithConfiguration,
+      Auth0NativeModule.hasValidAuth0InstanceWithConfiguration.bind(
+        Auth0NativeModule
+      ),
       clientId,
       domain
     );
@@ -55,7 +57,9 @@ export class NativeBridgeManager implements INativeBridge {
     // This is a new method we'd add to the native side to ensure the
     // underlying Auth0.swift/Auth0.android SDKs are configured.
     return this.a0_call(
-      Auth0NativeModule.initializeAuth0WithConfiguration,
+      Auth0NativeModule.initializeAuth0WithConfiguration.bind(
+        Auth0NativeModule
+      ),
       clientId,
       domain,
       localAuthenticationOptions,
@@ -64,7 +68,9 @@ export class NativeBridgeManager implements INativeBridge {
   }
 
   getBundleIdentifier(): Promise<string> {
-    return this.a0_call(Auth0NativeModule.getBundleIdentifier);
+    return this.a0_call(
+      Auth0NativeModule.getBundleIdentifier.bind(Auth0NativeModule)
+    );
   }
 
   async authorize(
@@ -79,7 +85,7 @@ export class NativeBridgeManager implements INativeBridge {
     const scheme =
       parameters.redirectUrl?.split('://')[0] ?? options.customScheme;
     const credential = await this.a0_call(
-      Auth0NativeModule.webAuth,
+      Auth0NativeModule.webAuth.bind(Auth0NativeModule),
       scheme,
       parameters.redirectUrl,
       parameters.state,
@@ -104,7 +110,7 @@ export class NativeBridgeManager implements INativeBridge {
     options: NativeClearSessionOptions
   ): Promise<void> {
     return this.a0_call(
-      Auth0NativeModule.webAuthLogout,
+      Auth0NativeModule.webAuthLogout.bind(Auth0NativeModule),
       options.customScheme,
       parameters.federated ?? false,
       parameters.returnToUrl
@@ -112,11 +118,16 @@ export class NativeBridgeManager implements INativeBridge {
   }
 
   async cancelWebAuth(): Promise<void> {
-    return this.a0_call(Auth0NativeModule.cancelWebAuth);
+    return this.a0_call(
+      Auth0NativeModule.cancelWebAuth.bind(Auth0NativeModule)
+    );
   }
 
   async saveCredentials(credentials: Credentials): Promise<void> {
-    return this.a0_call(Auth0NativeModule.saveCredentials, credentials);
+    return this.a0_call(
+      Auth0NativeModule.saveCredentials.bind(Auth0NativeModule),
+      credentials
+    );
   }
 
   async getCredentials(
@@ -128,7 +139,7 @@ export class NativeBridgeManager implements INativeBridge {
     // Assuming the native side can take an empty object for parameters.
     const params = parameters ?? {};
     return this.a0_call(
-      Auth0NativeModule.getCredentials,
+      Auth0NativeModule.getCredentials.bind(Auth0NativeModule),
       scope,
       minTtl ?? 0,
       params,
@@ -137,22 +148,30 @@ export class NativeBridgeManager implements INativeBridge {
   }
 
   async hasValidCredentials(minTtl?: number): Promise<boolean> {
-    return this.a0_call(Auth0NativeModule.hasValidCredentials, minTtl ?? 0);
+    return this.a0_call(
+      Auth0NativeModule.hasValidCredentials.bind(Auth0NativeModule),
+      minTtl ?? 0
+    );
   }
 
   async clearCredentials(): Promise<void> {
-    return this.a0_call(Auth0NativeModule.clearCredentials);
+    return this.a0_call(
+      Auth0NativeModule.clearCredentials.bind(Auth0NativeModule)
+    );
   }
 
   async resumeWebAuth(url: string): Promise<void> {
-    return this.a0_call(Auth0NativeModule.resumeWebAuth, url);
+    return this.a0_call(
+      Auth0NativeModule.resumeWebAuth.bind(Auth0NativeModule),
+      url
+    );
   }
 
   async getDPoPHeaders(
     params: DPoPHeadersParams
   ): Promise<Record<string, string>> {
     return this.a0_call(
-      Auth0NativeModule.getDPoPHeaders,
+      Auth0NativeModule.getDPoPHeaders.bind(Auth0NativeModule),
       params.url,
       params.method,
       params.accessToken,
@@ -162,6 +181,6 @@ export class NativeBridgeManager implements INativeBridge {
   }
 
   async clearDPoPKey(): Promise<void> {
-    return this.a0_call(Auth0NativeModule.clearDPoPKey);
+    return this.a0_call(Auth0NativeModule.clearDPoPKey.bind(Auth0NativeModule));
   }
 }

